### PR TITLE
AL/MIN-110. Balancing liquidation pools

### PR DESCRIPTION
**Pull request by lut1k**
Description:

1. Connected a new DEX module to the runtime.
2. In the DEX module has implemented a temporary function for an exchange at a rate of 1: 1.
3. Added oracle prices to genesis block.
4. Implemented a balancing algorithm.
5. Implemented tests for the above functionality.

**Changes in storage:**

There have been changes in the liquidation pool pallet:
StorageMap:
CurrencyId ->  "LiquidationPoolData": {
                                "deviation_threshold": "Rate",
                                "balance_ratio": "Rate"
                          }
StorageValue: BalancingPeriod

**Test scenarios/Use cases:**

If applicable, describe new use cases and leave @DenisRomanovsky  mention. 

The info will go to the Protocol documentation.

### CODE REVIEW

File: pallets/liquidation-pools/src/lib.rs [hunk 1]
```diff
@@ -269,213 +244,257 @@ pub mod module {
...
+		)?;
+
+		// Calculate sum_extra and sum_shortfall for all pools.
+		let (mut sum_oversupply, mut sum_shortfall) = information_vec.iter().try_fold(
+			(Balance::zero(), Balance::zero()),
+			|mut acc, pool| -> sp_std::result::Result<(Balance, Balance), DispatchError> {
+				let deviation_threshold = Self::liquidation_pools_data(pool.pool_id).deviation_threshold;
+
+				// right_border = ideal_balance + ideal_balance * deviation_threshold
+				let right_border = Rate::from_inner(pool.ideal_balance)
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs:388]: In my opinion, everything is correct:
line 391: .and_then(|v| v.checked_add(pool.ideal_balance))

File: pallets/liquidation-pools/src/lib.rs [hunk 2]
```diff
@@ -269,213 +244,257 @@ pub mod module {
...
+						.ok_or(Error::<T>::InvalidFeedPrice)?;
+					(
+						Rate::from_inner(max_shortfall)
+							.checked_div(&oracle_price)
+							.map(|x| x.into_inner())
+							.ok_or(Error::<T>::NumOverflow)?,
+						max_shortfall,
+					)
+				}
+			};
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs:459]: Cool remark. Fixed.

File: pallets/liquidation-pools/src/lib.rs [hunk 3]
```diff
@@ -269,213 +244,257 @@ pub mod module {
...
+				balance: information_vec[max_oversupply_index]
+					.balance
+					.checked_sub(bite_in_usd)
+					.ok_or(Error::<T>::NumOverflow)?,
+				oversupply: information_vec[max_oversupply_index]
+					.oversupply
+					.checked_sub(bite_in_usd)
+					.ok_or(Error::<T>::NumOverflow)?,
+				..information_vec[max_oversupply_index]
+			};
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs:479]: Fixed

File: pallets/liquidation-pools/src/lib.rs [hunk 4]
```diff
@@ -269,213 +244,257 @@ pub mod module {
...
+				balance: information_vec[max_shortfall_index]
+					.balance
+					.checked_add(bite_in_usd)
+					.ok_or(Error::<T>::NumOverflow)?,
+				shortfall: information_vec[max_shortfall_index]
+					.shortfall
+					.checked_sub(bite_in_usd)
+					.ok_or(Error::<T>::NumOverflow)?,
+				..information_vec[max_shortfall_index]
+			};
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs:491]: Fixed

File: pallets/liquidation-pools/src/lib.rs [hunk 5]
```diff
@@ -269,213 +244,257 @@ pub mod module {
...
+					ideal_balance,
+					oversupply,
+					shortfall,
+				});
+				Ok(acc)
+			},
+		)?;
+
+		// Calculate sum_extra and sum_shortfall for all pools.
+		let (mut sum_oversupply, mut sum_shortfall) = information_vec.iter().try_fold(
```
[the remark this answers is not in the document]
Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs:382]: Fixed.

File: pallets/liquidation-pools/src/lib.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (dev-vadim) [reply pallets/liquidation-pools/src/lib.rs]: My bad.

File: pallets/traits/src/lib.rs [hunk 6]
```diff
@@ -39,3 +40,21 @@ pub trait PriceProvider<CurrencyId> {
...
+pub trait DEXManager<AccountId, CurrencyId, Balance> {
+	fn swap_with_exact_supply(
+		who: &AccountId,
+		target_currency_id: CurrencyId,
+		supply_currency_id: CurrencyId,
+		supply_amount: Balance,
+		min_target_amount: Balance,
+	) -> sp_std::result::Result<Balance, DispatchError>;
+
+	fn swap_with_exact_target(
```
Comment (Apollo50) [inline pallets/traits/src/lib.rs:53]: You have to fix function signature ( swap two parameters: target_currency_id && supply_currency_id ).
This collision might make misunderstanding.
The same fn in dex pallet has got different signature: 
     ```
  
```
	fn swap_with_exact_target(
		who: &T::AccountId,
		supply_currency_id: CurrencyId,
		target_currency_id: CurrencyId,
		target_amount: Balance,
		max_supply_amount: Balance,
	) -> sp_std::result::Result<Balance, DispatchError> ;

File: pallets/traits/src/lib.rs [hunk 7]
```diff
@@ -39,3 +40,21 @@ pub trait PriceProvider<CurrencyId> {
 	fn lock_price(currency_id: CurrencyId);
 	fn unlock_price(currency_id: CurrencyId);
 }
+
+pub trait DEXManager<AccountId, CurrencyId, Balance> {
+	fn swap_with_exact_supply(
```
Comment (Apollo50) [inline pallets/traits/src/lib.rs:45]: the same

**Review by Apollo50 — COMMENTED** [review]
*

File: pallets/traits/src/lib.rs (hunk 6 above)
Comment (Apollo50) [inline pallets/traits/src/lib.rs:53]: You have to swap method signature ( target_currency_id && supply_currency_id )
This collision might make misunderstanding.

Comment (lut1k) [reply pallets/traits/src/lib.rs:53]: Fixed

**Review by Apollo50 — APPROVED** [review]

**Pull request merged**